python/pfs/utils/optimize_black_dots.py

```diff
@@ -0,0 +1,482 @@
+
+                # create prediction for where the points should be, when you do not see points
+                # if you do not see points from both sides of the black dots, do simpler extrapolation
+                # (np.sum(np.diff(idx_1)) == 1)
+                # if you see points from both sides of the black dots, do more complex interpolation
+                # (np.sum(np.diff(idx_1)) != 1)
+                if type_of_run == 'theta':
+                    if np.sum(np.diff(idx)) == 1:
+                        deg2_fit_to_single_point_y = np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                                                y_measurments[idx], 3)
+                        deg2_fit_to_single_point_x = np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                                                x_measurments[idx], 3)
+                    else:
+                        deg2_fit_to_single_point_y = np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                                                y_measurments[idx], 4)
+                        deg2_fit_to_single_point_x = np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                                                x_measurments[idx], 4)
+                else:
+                    if np.sum(np.diff(idx)) == 1:
+                        with warnings.catch_warnings():
+                            warnings.simplefilter("ignore", category=RuntimeWarning)
+                            deg2_fit_to_single_point_y, residuals_y =\
+                                np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                           y_measurments[idx], 1, full=True)[0:2]
+                            deg2_fit_to_single_point_x, residuals_x =\
+                                np.polyfit(np.arange(0, len(x_measurments))[idx],
+                                           x_measurments[idx], 1, full=True)[0:2]
+                    else:
+                        deg2_fit_to_single_point_y, residuals_y =\
+                            np.polyfit(np.arange(0, len(y_measurments))[idx],
+                                       y_measurments[idx], 3, full=True)[0:2]
+                        deg2_fit_to_single_point_x, residuals_x =\
+                            np.polyfit(np.arange(0, len(x_measurments))[idx],
+                                       x_measurments[idx], 3, full=True)[0:2]
+
+                    if len(residuals_y) == 0:
+                        residuals_y = np.array([99])
+                    if len(residuals_x) == 0:
+                        residuals_x = np.array([99])
+
+                poly_deg_x = np.poly1d(deg2_fit_to_single_point_x)
+                poly_deg_y = np.poly1d(deg2_fit_to_single_point_y)
+
+                # some cleaning, to exclude the results where the interpolation results make no sense
+                if type_of_run == 'theta':
+                    if np.sum(~idx) > 5:
```

Comment:
Why is this criterion used? Was it determined empirically?

Reply:
Basically yes. When we see only 5 points that is because the cobra is not moving, and as such there is no useful information to get from this fiber. Number 5 is somewhat arbitrary -> a small enough number that no good cobras are disregarded even if we take only a limited number of observations.